addons/l10n_pk_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,454 @@
+                    "success": False,
+                    "error": error_response,
+                    "blocking_level": "error",
+                }
+            }
+        json_dump = json.dumps(response)
+        json_name = "%s_einvoice.json" % (invoice.name.replace("/", "_"))
+        attachment = self.env["ir.attachment"].create({
+            "name": json_name,
+            "raw": json_dump.encode(),
+            "res_model": "account.move",
+            "res_id": invoice.id,
+            "mimetype": "application/json",
+        })
+        return {
+            invoice: {
+                "success": True,
+                "attachment": attachment
+            }
+        }
+
+    def _get_l10n_pk_total_tax_values(self, tax_lines, move):
+        sales_tax_total = 0.0
+        non_sales_tax_total = 0.0
+        sale_group_ids = (
+            self._get_l10n_pk_sales_groups(move.company_id)
+            + self._get_l10n_pk_withhold_groups(move.company_id)
+        ).ids
+        for tax_group_id in tax_lines:
+            if tax_group_id in sale_group_ids:
+                sales_tax_total += tax_lines[tax_group_id]['amount']
+            else:
+                non_sales_tax_total += tax_lines[tax_group_id]['amount']
+        return sales_tax_total, non_sales_tax_total
+
+    def _get_l10n_pk_tax_details_by_line(self, tax_details):
+        l10n_pk_tax_details = {}
+        for tax_detail in tax_details.values():
+            line_code = tax_detail["tax"].tax_group_id.id
+            l10n_pk_tax_detail = l10n_pk_tax_details.get(line_code, {})
+            l10n_pk_tax_detail.update(dict.fromkeys(("rate", "amount", "amount_currency"), 0.0))
+            l10n_pk_tax_detail["rate"] += tax_detail["tax"].amount
+            l10n_pk_tax_detail["amount"] += tax_detail["tax_amount"]
+            l10n_pk_tax_detail["amount_currency"] += tax_detail["tax_amount_currency"]
+            l10n_pk_tax_details[line_code] = l10n_pk_tax_detail
+        return l10n_pk_tax_details
+
+    # ================================ API methods ===========================
+
+    @api.model
+    def _l10n_pk_edi_connect_to_server(self, company, payload, use_timeout=False):
+        proxy_address = company.sudo().l10n_pk_edi_proxy
+        if not proxy_address:
+            raise UserError(_("Please add a proxy server from configuration to test APIs"))
+        try:
+            response = requests.post(
+                DEFAULT_ENDPOINT,
+                json=payload,
+                headers={
+                    'Authorization': f"Bearer {company.sudo().l10n_pk_edi_token}",
+                    'Content-Type': 'application/json',
+                },
+                proxies={
+                    'http': 'http://' + proxy_address,
+                    'https': 'http://' + proxy_address,
+                },
+                # While using proxy server it may take longer than 25 sec
+                timeout=self.env["ir.config_parameter"].sudo().get_param(
```

Comment:
not sure that is acceptable security-wise

Reply:
We have no option either give increased value i.e. `60` or give user to increase the timeout. The IP should be registered with government and when we might take it this IAP we might need it. Because currently we used a proxy server. `25` sec seems not to be enough all the times, currently out 3 of 5 requests, 2 requests seems to be timeout while using the proxy.

Reply:
@hamo-odoo Is there a way to recuperate the invoice afterwards when it times out?  Normally, the proxy server should not have much influence on it...  (is the proxy needed?)

Reply:
@jco-odoo Yes the proxy is needed. The IP should be static and the API is only accessible from Pakistan. If we go for IAP. We might need to set up a local server that communicates with the API. `User -> IAP -> Local Proxy Server`. This would be require a good amount of timeout.